### PR TITLE
Chapter 10: Manage User

### PR DESCRIPTION
**Pull request by hazwanizafirah**
ready

### CODE REVIEW

**Comment by hazwanizafirah** [comment]
image
Ready but I have this problem. Can I know how to fix this problem?

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
 
-  def show; end
+
+  def index
+    @users = User.paginate(page: params[:page])
+  end
+
+  def show
+    @user = User.find_by(id: params[id:])
+   end
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:13]: align end with def

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -17,17 +26,53 @@ def create
...
+      store_location
+      flash[:danger] = t("Please log in.")
+      redirect_to login_url
+    end
+  end
+
+  def correct_user
+    @user = User.find_by(id: params[:id])
+    return if current_user?(@user)
+    flash[danger:] = t("You are not authorized")
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:70]:
```suggestion
    return if current_user?(@user)
    
    flash[danger:] = t("You are not authorized")
```

File: config/routes.rb [hunk 3]
```diff
@@ -7,10 +7,14 @@
...
-    resources :users, only: %i(new create show)
+    resources :users, only: %i(show new create destroy edit update)
 
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     get "/logout", to: "sessions#destroy"
+
+    get "/edit", to: "users#edit"
+    post "/edit", to: "users#update"
+    get "/users", to: "users#index"
```
Comment (anlvq-1058) [inline config/routes.rb:18]: what do you know about resources line 10?
why do you need to create more routes to users_controller?

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
image
I still got this error. how to fix?

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,7 +1,16 @@
...
+  before_action :admin_user, only: :destroy
 
-  def show; end
+
+  def index
+    @users = User.paginate(page: params[:page])
+  end
+
+  def show
+    @user = User.find_by(id: params[id:])
```
Comment (thaodtd-0665) [inline app/controllers/users_controller.rb:12]: if you remove "before_action :load_user, only: :show" you need handle in this
```suggestion
    @user = User.find_by(id: params[id:])
    return if @user
    
    flash[:error] = t(:error)
    redirect_to root_path    
```

File: app/views/layouts/_header.html.erb [hunk 5]
```diff
@@ -7,15 +7,15 @@
       <li><%= link_to t(:help), help_path %></li>
 
       <% if logged_in? %>
-        <li><%= link_to t(:users), "#" %></li>
+        <li><%= link_to t(:users), users_path %></li>
```
Comment (thaodtd-0665) [inline app/views/layouts/_header.html.erb:10]:
```suggestion
        <li><%= link_to t(:users), users_path(current_user) %></li>
```

**Comment by hazwanizafirah** [comment]
ready

**Comment by hazwanizafirah** [comment]
image
I already follow your comment, but I still got the same error :(

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -17,17 +30,53 @@ def create
...
+    @user = User.find_by(id: params[:id])
+    return if current_user?(@user)
+    flash[danger:] = t("You are not authorized")
+    redirect_to(root_url) unless current_user?(@user)
   end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
+  end
+end
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb:81]: remove this line, it's redundant

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -17,17 +30,53 @@ def create
...
     params.require(:user).permit(:name, :email, :password, :password_confirmation)
   end
 
-  def load_user
-    @user = User.find_by id: params[:id]
-    return if @user
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t("Please log in.")
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb:66]: it's right? `t("Please log in.")` ==> `t "please_log_in"`

File: app/views/users/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, t("All users")) %>
+<h1><%= t("All users") %></h1>
+
+<%= will_paginate @users %>
+<ul class="users">
+  <% @users.each do |user| %>
+   <%= render @users %>
```
Comment (huongnt-2545) [inline app/views/users/index.html.erb:7]:
```suggestion
   <%= render user %>
```

File: config/routes.rb [hunk 9]
```diff
@@ -12,5 +12,6 @@
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     get "/logout", to: "sessions#destroy"
+
```
Comment (huongnt-2545) [inline config/routes.rb:15]: you missing define additional resources for user actions (index, edit, update, destroy) at line 10

**Comment by hazwanizafirah** [comment]
ready

File: config/routes.rb [hunk 10]
```diff
@@ -7,10 +7,14 @@
 
     get "/signup", to: "users#new"
     post "/signup", to: "users#create"
-    resources :users, only: %i(new create show)
+    resources :users, only: %i(show new create destroy edit update)
```
Comment (huongnt-2545) [inline config/routes.rb:10]: comprehensive actions
```suggestion
    resources :users
```

File: config/routes.rb [hunk 11]
```diff
@@ -7,10 +7,14 @@
...
 
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     get "/logout", to: "sessions#destroy"
+
+    get "/edit", to: "users#edit"
```
Comment (huongnt-2545) [inline config/routes.rb:16]: remove line 16 -> 18

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -17,17 +30,48 @@ def create
...
     params.require(:user).permit(:name, :email, :password, :password_confirmation)
   end
 
-  def load_user
-    @user = User.find_by id: params[:id]
-    return if @user
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t "Please log in."
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb:66]:
```suggestion
      flash[:danger] = t "please_log_in."
```

**Comment by hazwanizafirah** [comment]
image
Can I know how to fix this problem?

**Comment by hazwanizafirah** [comment]
ready

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -1,7 +1,10 @@
 module UsersHelper
-  def gravatar_for(user)
+  def gravatar_for(user, options = {size: 80})
     gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
-    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
-    image_tag(gravatar_url, alt: user.name, class: "gravatar")
+    size = options[:size]
+    gravatar_url =
+"https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
```
Comment (ngocvt-0484) [inline app/helpers/users_helper.rb:6]: wrong indent

File: config/routes.rb (hunk 11 above)
Comment (ngocvt-0484) [reply config/routes.rb:16]: not fix comment yet

**Comment by hazwanizafirah** [comment]
ready

File: config/routes.rb [hunk 14]
```diff
@@ -6,11 +6,11 @@
...
 
     get "/signup", to: "users#new"
-    post "/signup", to: "users#create"
-    resources :users, only: %i(new create show)
 
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     get "/logout", to: "sessions#destroy"
+
+    resources :users, except: %i(new)
```
Comment (huongnt-2545) [inline config/routes.rb:14]: because UsersController has comprehensive actions
```
resources :users
```

**Comment by hazwanizafirah** [comment]
ready

File: config/routes.rb [hunk 15]
```diff
@@ -6,11 +6,12 @@
     get 'static_pages/help', as: "help"
 
     get "/signup", to: "users#new"
-    post "/signup", to: "users#create"
```
Comment (huongnt-2545) [inline config/routes.rb:9]: keep this line

**Comment by hazwanizafirah** [comment]
ready

**Comment by hazwanizafirah** [comment]
ready

**Review by huongnt-2545 — APPROVED** [review]
merged
You can merge this pull

**Pull request merged**